File: hyejee/kakao2022/주차 요금 계산.swift
import Foundation

class Record {
    let inTime: Int
    var outTime: Int = (23 * 60 + 59)
    var accumulatedMinutes: Int {
        return self.outTime - self.inTime
    }
    
    init(inTime: Int) {
        self.inTime = inTime
    }
}

func parseRecord(of record: String) -> (Int, String, String)? {
    let strings: [String] = record.split(separator: " ").map({ String($0) })
    
    guard strings.count == 3 else {
        return nil
    }
    
    // 시간 계산
    let times: [Int] = strings[0].split(separator: ":").compactMap({ Int(String($0)) })
    guard times.count == 2 else {
        return nil
    }
    
    let minutes: Int = times[0] * 60 + times[1]
    
    return (minutes, strings[1], strings[2])
}

func calculateFee(of minutes: Int, _ fees: [Int]) -> Int {
    var extra: Int = 0
    
    if minutes -  fees[0] > 0 {
        extra = Int(ceil(Double(minutes - fees[0]) / Double(fees[2]))) * fees[3]
    }
    
    return fees[1] + extra
}

func solution(_ fees:[Int], _ records:[String]) -> [Int] {
    var _records: [String: [Record]] = [:]
    
    for record in records {
        guard let parsed: (Int, String, String) = parseRecord(of: record) else {
            continue
        }
        
        switch parsed.2 {
        case "IN":
            _records[parsed.1, default: []].append(Record(inTime: parsed.0))
        case "OUT":
            if let last = _records[parsed.1, default: []].last {
                last.outTime = parsed.0
            }
        default:
            return []
        }
    }
    
    var answer: [Int] = []
    
    for carNumber in _records.keys.sorted() {
        let minutes: Int = _records[carNumber, default: []].map({ $0.accumulatedMinutes }).reduce(into: 0) {
            $0 += $1
        }
        
        let fee: Int = calculateFee(of: minutes, fees)
        
        answer.append(fee)
    }
    
    return answer
}

File: seungmi/kakao2022/주차요금계산.py
import datetime as dt
import math

def cal_time(in_time,out_time):
    time = str(dt.datetime.strptime(out_time,'%H:%M')-dt.datetime.strptime(in_time,'%H:%M'))
    return int(time.split(":")[0])*60+int(time.split(":")[1])

def solution(fees, records):
    record_dic={}
    fee_dic={}

    #차량 별 사용시간 계산
    #record_dic={"차량번호":[입차시간, 입/출차 여부,누적시간] }
    for record in records:
        car_info=record.split(" ")
        if car_info[1] not in record_dic.keys():
            record_dic[car_info[1]]=[car_info[0],car_info[2],0]
            continue
        if car_info[2]=="IN":
            record_dic[car_info[1]][0]=car_info[0]
            record_dic[car_info[1]][1]=car_info[2]
            
        else:
            record_dic[car_info[1]][2]+=cal_time(record_dic[car_info[1]][0],car_info[0])
            record_dic[car_info[1]][1]=car_info[2]
            
    #미 출차 차량 시간계산 /사용 요금 계산
    for car_num in record_dic:
        if record_dic[car_num][1]=="IN":
            record_dic[car_num][2]+=cal_time(record_dic[car_num][0],"23:59")
            record_dic[car_num][1]="OUT"
        
        if record_dic[car_num][2]<fees[0]:
            fee_dic[car_num]=fees[1]
            continue
        fee_dic[car_num]=fees[1]+math.ceil((record_dic[car_num][2]-fees[0])/fees[2])*fees[3]
    
    return [i[1] for i in sorted(fee_dic.items())]